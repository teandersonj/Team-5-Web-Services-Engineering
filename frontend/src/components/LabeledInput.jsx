/**
 * LabeledInput component for forms, consisting of a label and a specified input
 * @param {props.id} ID - The ID of the input, used to identify it
 * -- Currently, the ID is also used as the name of the input
 * @param {props.label} Label - The label to display before the input
 * @param {props.type} Type - The type of input to display (text, email, password, etc.)
 * @param {props.value} DefaultValue - The defaultValue of the input
 * @param {props.orientation} Orientation - Horizontal: Display the label and input side by side. Vertical: Display the label above the input
 * @param {props.placeholder} Placeholder - The placeholder text to display in the input
 * @param {props.onChange} onChange - The function to call when the input changes
 * @param {props.containerStyle} containerStyle - The style to apply to the container div
 * @param {props.labelStyle} labelStyle - The style to apply to the label
 * @param {props.inputStyle} inputStyle - The style to apply to the input
 * @param {props.containerClassName} containerClassName - The className to apply to the container div
 * @param {props.labelClassName} labelClassName - The className to apply to the label
 * @param {props.inputClassName} inputClassName - The className to apply to the input
 * @returns <LabeledInput /> LabeledInput JSX component
 */
export default function LabeledInput(props) {
    const { id, label, type, defaultValue, orientation = "vertical", placeholder, onChange, containerStyle, labelStyle, inputStyle, containerClassName, labelClassName, inputClassName } = props;
    
    let orientationStyle = orientation === "horizontal" ? { display: "flex", flexDirection: "row", alignItems: "center" } : { display: "flex", flexDirection: "column", alignItems: "flex-start" };

    return (
        <div className={containerClassName} style={{ ...containerStyle, ...orientationStyle }}>
            <label htmlFor={id} className={labelClassName} style={labelStyle}>{label}</label>
            <input key={id} id={id} name={id} type={type} defaultValue={defaultValue} placeholder={placeholder} onChange={onChange} className={inputClassName} style={inputStyle} />
        </div>
    );
};
