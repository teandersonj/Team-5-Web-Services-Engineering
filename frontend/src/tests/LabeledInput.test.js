// Disable these linting rules since we want to access the DOM directly
/* eslint testing-library/no-node-access: 0 */
/* eslint testing-library/no-container: 0 */

import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import LabeledInput from "../components/LabeledInput";

describe('LabeledInput Component Tests', () => {
    // Use Jest and react-testing-library functions to test the component
    // First, does it render without any props?
    // Nex, given various props, does it alter the rendered output as expected?
    // Finally, does it call the correct function when the input is changed?
    // Test rendering without props
    it('Renders without props', () => {
        const { container } = render(<LabeledInput />);
        expect(container).toMatchSnapshot();
    });

    it("Renders with certain props (Labeled Email Required Input with Attributes)", () => {
        const { container } = render(<LabeledInput id="email" label="Email" type="email" defaultValue="test@test.com" disabled={true} required={true} orientation="horizontal" placeholder="Enter your email" />);
        expect(container).toMatchSnapshot();
        // Make sure the input has the correct properties 4
        const input = container.querySelector("input");
        expect(input).toHaveAttribute("id", "email");
        expect(input).toHaveAttribute("name", "email");
        expect(input).toHaveAttribute("type", "email");
        // defaultValue gets converted to value in the DOMdf
        expect(input).toHaveAttribute("value", "test@test.com");
        expect(input).toHaveAttribute("placeholder", "Enter your email");
        expect(input).toHaveAttribute("disabled");
        expect(input).toHaveAttribute("required");
    });

    it("Renders with certain props (Labeled Password Optional Input with Attributes)", () => {
        const { container } = render(<LabeledInput id="password" label="Password" type="password" defaultValue="password" disabled={false} required={false} orientation="vertical" placeholder="Enter your password" />);
        expect(container).toMatchSnapshot();
        // Make sure the input has the correct properties
        const input = container.querySelector("input");
        expect(input).toHaveAttribute("id", "password");
        expect(input).toHaveAttribute("name", "password");
        expect(input).toHaveAttribute("type", "password");
        // defaultValue gets converted to value in the DOM
        expect(input).toHaveAttribute("value", "password");
        expect(input).toHaveAttribute("placeholder", "Enter your password");
        expect(input).not.toHaveAttribute("disabled");
        expect(input).not.toHaveAttribute("required");
    });

    // Let's pass a dummy function and see if it gets called when the input is changed
    it("Calls the correct function when the input is changed", async () => {
        const user = userEvent.setup();
        const mockOnChange = jest.fn();
        const { container } = render(<LabeledInput id="email" label="Email" type="email" defaultValue="" disabled={false} required={false} orientation="vertical" placeholder="Enter your email" onChange={(event) => { mockOnChange(event.target.value) }} />);
        const input = container.querySelector("input");
        // Simulate typing into the input
        await user.type(input, "T");
        // Make sure the function was called the appropriate number of times
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        // Make sure the function was called with the correct arguments
        expect(mockOnChange).toHaveBeenCalledWith("T");
    });
});
