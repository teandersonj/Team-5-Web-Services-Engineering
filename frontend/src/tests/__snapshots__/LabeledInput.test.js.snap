// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabeledInput Component Tests Renders with certain props (Labeled Email Required Input with Attributes) 1`] = `
<div>
  <div
    style="display: flex; flex-direction: row; align-items: center;"
  >
    <label
      for="email"
    >
      Email
    </label>
    <input
      disabled=""
      id="email"
      name="email"
      placeholder="Enter your email"
      required=""
      type="email"
      value="test@test.com"
    />
  </div>
</div>
`;

exports[`LabeledInput Component Tests Renders with certain props (Labeled Password Optional Input with Attributes) 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column; align-items: flex-start;"
  >
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      placeholder="Enter your password"
      type="password"
      value="password"
    />
  </div>
</div>
`;

exports[`LabeledInput Component Tests Renders without props 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column; align-items: flex-start;"
  >
    <label>
      Input
    </label>
    <input
      type="text"
      value=""
    />
  </div>
</div>
`;
